libsolutil/IpfsHash.cpp

```diff
@@ -53,36 +53,127 @@ string base58Encode(bytes const& _data)
 	reverse(output.begin(), output.end());
 	return output;
 }
+struct Chunk
+{
+	Chunk() = default;
+	Chunk(bytes _hash, size_t _size, size_t _blockSize):
+		hash(std::move(_hash)),
+		size(_size),
+		blockSize(_blockSize)
+	{}
+	bytes hash;
+	size_t size;
+	size_t blockSize;
+};
+
+using Chunks = vector<Chunk>;
+
+/// Builds a tree starting from the bottom level where nodes are data nodes.
+/// Data nodes should be calculated and passed as the only level in chunk levels
+/// Each next level is calculated as following:
+///   - Pick up to maxChildNum (174) nodes until a whole level is added, group them and pass to the node in the next level
+///   - Do this until the current level has only one node, return the hash in that node
+bytes groupChunksBottomUp(vector<Chunks>& chunkLevels)
```

Comment:
```suggestion
bytes groupChunksBottomUp(vector<Chunks>& _chunkLevels)
```

Maybe it even makes sense to even use `groupChunksBottomUp(Chunks _allChunks)`

and then do

```
vector<Chunks> chunkLevels;
chunkLevels.front() = std::move(_allChunks);
```

Reply:
I am not sure if I follow this one. How would we call the function in this case?

Reply:
I see now